Sources/SotoCore/Credential/ConfigFileLoader.swift

```diff
@@ -0,0 +1,207 @@
+        case ecsContainer = "EcsContainer"
+    }
+
+    /// Errors occurring when loading credentials and profile configuration
+    ///
+    /// - missingProfile: If the profile requested was not found
+    /// - missingAccessKeyId: If the access key ID was not found
+    /// - missingSecretAccessKey: If the secret access key was not found
+    enum ConfigFileError: Error, Equatable {
+        case invalidCredentialFileSyntax
+        case missingProfile(String)
+        case missingAccessKeyId
+        case missingSecretAccessKey
+        case missingSourceProfile
+    }
+
+    /// Load shared credentials and profile configuration from passed-in byte-buffers
+    ///
+    /// Credentials file settings have precedence over profile configuration settings
+    /// https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-precedence
+    ///
+    /// - Parameters:
+    ///   - credentialsByteBuffer: contents of AWS shared credentials file (usually `~/.aws/credentials`
+    ///   - configByteBuffer: contents of AWS profile configuration file (usually `~/.aws/config`
+    ///   - profile: named profile to load (usually `default`)
+    ///   - context: credential provider factory context
+    /// - Returns: Credential Provider (StaticCredentials or STSAssumeRole)
+    static func sharedCredentials(from credentialsByteBuffer: ByteBuffer,
+                                  configByteBuffer: ByteBuffer? = nil,
+                                  for profile: String,
+                                  context: CredentialProviderFactory.Context) throws -> CredentialProvider {
+        var config: ConfigFileLoader.ProfileConfig?
+        if let byteBuffer = configByteBuffer {
+            config = try loadProfileConfig(from: byteBuffer, for: profile)
+        }
+        let credentials = try loadCredentials(from: credentialsByteBuffer, for: profile, sourceProfile: config?.sourceProfile)
+
+        // When `role_arn` is defined, temporary credentials must be loaded via STS Assume Role operation
+        if let roleArn = credentials.roleArn ?? config?.roleArn {
+            // If `role_arn` is defined, `source_profile` must be defined too (don't yet support `credential_source`).
+            guard credentials.sourceProfile != nil || config?.sourceProfile != nil else {
+                throw ConfigFileError.missingSourceProfile
+            }
+            // Proceed with STSAssumeRole operation
+            let sessionName = credentials.roleSessionName ?? config?.roleSessionName ?? UUID().uuidString
+            let request = STSAssumeRoleRequest(roleArn: roleArn, roleSessionName: sessionName)
+            let region = config?.region ?? .useast1
+            let provider = STSAssumeRoleCredentialProvider(request: request, credentialProvider: .default, region: region, httpClient: context.httpClient)
+            return RotatingCredentialProvider(context: context, provider: provider)
```

Comment:
Should the `credentialProvider` here, be either the static credentials in the `profile` or the equivalent provider for the `credentialSource`

If you use the default chain it could load credentials for the assume role via ConfigFileCredentialProvider again. And if the default chain includes the ConfigFile it would create recursive credential loading.

Reply:
Good point. Will update to use `StaticCredentials` for the STS operation.